### PR TITLE
 Longest Increasing subsequence using binary search most optimal approach (Modified)

### PR DESCRIPTION
**Pull request by namanmodi65**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

**Review by realstealthninja — CHANGES_REQUESTED** [review]
Please use the format of the header comment used in the contributing.md file

**Comment by namanmodi65** [comment]
I did not get what you are trying to say can you please elaborate

File: search/Longest_Increasing_Subsequence_using_binary_search.cpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
+#include <cassert>   /// for std::assert
```
Comment (realstealthninja) [inline search/Longest_Increasing_Subsequence_using_binary_search.cpp:1]: needs a header comment explaining the algorithm, if relevant math formulas in latex format, authors name

this is the format;
```cpp
/**
 * @file
 * @brief Add one-line description here. Should contain a Wikipedia
 * link or another source explaining the algorithm/implementation.
 * @details
 * This is a multi-line
 * description containing links, references,
 * math equations, etc.
 * @author [Name](https://github.com/handle)
 * @see related_file.cpp, another_file.cpp
 */
 ```

**Review by realstealthninja — CHANGES_REQUESTED** [review]

**Comment by realstealthninja** [comment]
is this for hacktoberfest?

**Comment by namanmodi65** [comment]
Yes this is for hacktoberfest

File: search/Longest_Increasing_Subsequence_using_binary_search.cpp (hunk 1 above)
Comment (namanmodi65) [reply search/Longest_Increasing_Subsequence_using_binary_search.cpp:52]: done

File: search/Longest_Increasing_Subsequence_using_binary_search.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
+
+/**
+ * @file
+ * @brief find the length of the Longest Increasing Subsequence (LIS)
+ * using Binary Search(https://en.wikipedia.org/wiki/Longest_increasing_subsequence)
```
Comment (realstealthninja) [inline search/Longest_Increasing_Subsequence_using_binary_search.cpp:5]:
```suggestion
 * using [Binary Search](https://en.wikipedia.org/wiki/Longest_increasing_subsequence)
```

File: search/Longest_Increasing_Subsequence_using_binary_search.cpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+/**
+ * @brief Test cases for Longest Increasing Subsequence function
+ * @returns void
+ */
+static void tests() {
+    std::vector<int> arr = {10, 9, 2, 5, 3, 7, 101, 18};
+    assert(longest_increasing_subsequence_using_binary_search(arr) == 4);
+
+    std::vector<int> arr2 = {0, 1, 0, 3, 2, 3};
```
Comment (realstealthninja) [inline search/Longest_Increasing_Subsequence_using_binary_search.cpp:87]: add tests for negative numbers and empty array

File: search/Longest_Increasing_Subsequence_using_binary_search.cpp (hunk 2 above)
Comment (namanmodi65) [reply search/Longest_Increasing_Subsequence_using_binary_search.cpp]: Done

**Comment by realstealthninja** [comment]
Why is the pull request called longest increasing sequence but it is an implementation of floyd warshall?

**Comment by namanmodi65** [comment]
> Why is the pull request called longest increasing sequence but it is an implementation of floyd warshall?

now its correct

**Comment by namanmodi65** [comment]
> Why is the pull request called longest increasing sequence but it is an implementation of floyd warshall?

By mistake i have pushed the floyd warshall code with longest increasing sequence now i have removed that code.

File: search/Longest_Increasing_Subsequence_using_binary_search.cpp [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+#include <vector>    /// for std::vector
+#include <algorithm> /// for std::lower_bound
+
+/**
+ * @brief Function to find the length of the Longest Increasing Subsequence (LIS)
+ * using Binary Search
+ * @param nums The input vector of integers
+ * @return The length of the longest increasing subsequence
+ */
+int longest_increasing_subsequence_using_binary_search(std::vector<int>& nums) {
```
Comment (realstealthninja) [inline search/Longest_Increasing_Subsequence_using_binary_search.cpp:63]: since subsequence could be any type that is comparable you could make this function generic

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: search/Longest_Increasing_Subsequence_using_binary_search.cpp (hunk 4 above)
Comment (namanmodi65) [reply search/Longest_Increasing_Subsequence_using_binary_search.cpp]: Done

[the remark this answers is not in the document]
Comment (namanmodi65) [reply search/Longest_Increasing_Subsequence_using_binary_search.cpp]: Done

File: search/Longest_Increasing_Subsequence_using_binary_search.cpp [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+#include <cstdint>   /// for std::uint32_t
+
+/**
+ * @brief Function to find the length of the Longest Increasing Subsequence (LIS)
+ * using Binary Search
+ * @tparam T The type of the elements in the input vector
+ * @param nums The input vector of elements of type T
+ * @return The length of the longest increasing subsequence
+ */
+
```
Comment (realstealthninja) [inline search/Longest_Increasing_Subsequence_using_binary_search.cpp:64]:
```suggestion
```

File: search/Longest_Increasing_Subsequence_using_binary_search.cpp [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    std::cout << "All tests have successfully passed!\n";
+}
+
+/**
+ * @brief Main function to run tests
+ * @returns 0 on exit
+ */
+int main() {
+    tests();
```
Comment (realstealthninja) [inline search/Longest_Increasing_Subsequence_using_binary_search.cpp:116]:
```suggestion
    tests(); // run self test implementation
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: search/Longest_Increasing_Subsequence_using_binary_search.cpp (hunk 6 above)
Comment (namanmodi65) [reply search/Longest_Increasing_Subsequence_using_binary_search.cpp]: Done

**Review by realstealthninja — APPROVED** [review]
Lgtm, thanks for this contribution
